Description: Enlarge too short target string for sprintf
Bug-Debian: https://bugs.debian.org/897864
Forwarded: https://github.com/KorfLab/SNAP/issues/3
Author: Andreas Tille <tille@debian.org>
Last-Update:  Mon, 07 May 2018 23:38:54 +0200

--- a/snap.c
+++ b/snap.c
@@ -84,7 +84,7 @@ int main (int argc, char *argv[]) {
 	zoeCDS          gene;
 	zoeVec          genes;
 	int             label, i;
-	char            option[16], name[16];
+	char            option[20], name[16];
 	FILE          * aa_stream = NULL;
 	FILE          * tx_stream = NULL;
 	FILE          * xd_stream  = NULL;
